pkg/volumes/alicloud/discovery.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alicloud
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/golang/glog"
+	"kope.io/etcd-manager/pkg/privateapi/discovery"
+	"kope.io/etcd-manager/pkg/volumes"
+)
+
+// AlicloudVolumes allows us to discover our peer nodes
+var _ discovery.Interface = &AlicloudVolumes{}
+
+func (a *AlicloudVolumes) Poll() (map[string]discovery.Node, error) {
+	nodes := make(map[string]discovery.Node)
+
+	allVolumes, err := a.findVolumes(false)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceToVolumeMap := make(map[string]*volumes.Volume)
+	for _, v := range allVolumes {
+		if v.AttachedTo != "" {
+			instanceToVolumeMap[v.AttachedTo] = v
+		}
+	}
+
+	if len(instanceToVolumeMap) != 0 {
+		request := ecs.CreateDescribeInstancesRequest()
+		request.RegionId = a.region
+		request.PageSize = requests.NewInteger(PageSizeLarge)
+
+		var ids []string
+		for id := range instanceToVolumeMap {
+			ids = append(ids, id)
+		}
+		request.InstanceIds = "[\"" + strings.Join(ids, "\",\"") + "\"]"
+
+		response, err := a.ecs.DescribeInstances(request)
+		if err != nil {
+			return nil, fmt.Errorf("error from Alicloud DescribeInstances: %v", err)
+		}
+
+		for _, instance := range response.Instances.Instance {
```

Comment:
what happens if same instance do have 2 volumes? Like in kops we have events and main etcds in same machine. Does response.Instances.Instance contain duplicate instances or only unique? If unique, this logic is not going to work

Reply:
The response.Instances.Instance contains only unique instances, but that is also fine. Because there won't be two volumes with the same tags in one instance. 

The `allVolumes, err := a.findVolumes(false)` above is already filtered by tags, so I don't think there will be two volumes with same tags attached to the same instances. 

I also checked `aws/discovery.go`, basically they are with the same logic here.